File: backend/app/models/document.py
"""Document / chunk / job models."""

from __future__ import annotations

from datetime import datetime

from sqlalchemy import BigInteger, DateTime, ForeignKey, Integer, String, func
from sqlalchemy.dialects.mysql import MEDIUMTEXT
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base
from app.models.base import TimestampMixin, UUIDPrimaryKeyMixin


class Document(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "document"

    org_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("organization.id"), nullable=False, index=True
    )
    kb_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("knowledge_base.id"), nullable=False, index=True
    )
    file_name: Mapped[str] = mapped_column(String(255), nullable=False)
    storage_path: Mapped[str] = mapped_column(String(512), nullable=False)
    file_hash: Mapped[str] = mapped_column(String(64), nullable=False)
    file_type: Mapped[str] = mapped_column(String(32), nullable=False)
    file_size: Mapped[int] = mapped_column(BigInteger, nullable=False)
    page_count: Mapped[int | None] = mapped_column(Integer)
    status: Mapped[str] = mapped_column(String(32), nullable=False, default="UPLOADED")
    created_by: Mapped[str] = mapped_column(
        String(36), ForeignKey("sys_user.id"), nullable=False, index=True
    )

    jobs: Mapped[list[DocumentJob]] = relationship(
        back_populates="document",
        cascade="all, delete-orphan",
        lazy="selectin",
    )
    chunks: Mapped[list[DocumentChunk]] = relationship(
        back_populates="document",
        cascade="all, delete-orphan",
        lazy="selectin",
    )


class DocumentChunk(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "document_chunk"

    org_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("organization.id"), nullable=False, index=True
    )
    kb_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("knowledge_base.id"), nullable=False, index=True
    )
    document_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("document.id", ondelete="CASCADE"), nullable=False
    )
    chunk_index: Mapped[int] = mapped_column(Integer, nullable=False)
    content: Mapped[str] = mapped_column(MEDIUMTEXT, nullable=False)
    content_hash: Mapped[str] = mapped_column(String(64), nullable=False)
    page: Mapped[int | None] = mapped_column(Integer)
    section: Mapped[str | None] = mapped_column(String(255))
    token_count: Mapped[int | None] = mapped_column(Integer)

    document: Mapped[Document] = relationship(back_populates="chunks")


class DocumentJob(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "document_job"

    document_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("document.id", ondelete="CASCADE"), nullable=False
    )
    job_type: Mapped[str] = mapped_column(String(64), nullable=False, default="PARSE_INDEX")
    status: Mapped[str] = mapped_column(String(32), nullable=False, default="PENDING")
    progress: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    retry_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    error_code: Mapped[str | None] = mapped_column(String(64))
    error_message: Mapped[str | None] = mapped_column(String(1024))
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=False))
    finished_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=False))

    document: Mapped[Document] = relationship(back_populates="jobs")

File: frontend/apps/web-ele/src/types/document.ts
/** Document types aligned with backend DTO */

export type DocumentStatus =
  | 'UPLOADED'
  | 'PROCESSING'
  | 'READY'
  | 'FAILED'
  | 'DELETED';

export interface DocumentJobBrief {
  id: string;
  status: string;
  progress: number;
  error_code: string | null;
  error_message: string | null;
}

export interface DocumentItem {
  id: string;
  org_id: string;
  kb_id: string;
  file_name: string;
  file_hash: string;
  file_type: string;
  file_size: number;
  page_count: number | null;
  status: DocumentStatus;
  created_by: string;
  created_at: string | null;
  updated_at: string | null;
  job_id?: string;
  latest_job?: DocumentJobBrief | null;
}

export interface DocumentPage {
  items: DocumentItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface DocumentJob {
  id: string;
  document_id: string;
  job_type: string;
  status: string;
  progress: number;
  retry_count: number;
  error_code: string | null;
  error_message: string | null;
  started_at: string | null;
  finished_at: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export const DOCUMENT_STATUS_MAP: Record<
  Exclude<DocumentStatus, 'DELETED'>,
  { label: string; type: 'info' | 'warning' | 'success' | 'danger' }
> = {
  UPLOADED: { label: '已上传', type: 'info' },
  PROCESSING: { label: '解析中', type: 'warning' },
  READY: { label: '已就绪', type: 'success' },
  FAILED: { label: '解析失败', type: 'danger' },
};
